kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/storage/AwsS3DirectoryLister.scala

```diff
@@ -0,0 +1,140 @@
+
+import java.time.Clock
+import scala.collection.mutable
+import scala.jdk.CollectionConverters.IteratorHasAsScala
+import scala.jdk.CollectionConverters.ListHasAsScala
+
+abstract class AwsS3DirectoryLister(
+  implicit
+  connectorTask: ConnectorTaskId,
+  s3Client:      S3Client,
+) extends StorageInterface {
+  def findDirectories(
+    bucketAndPrefix:  RemoteS3RootLocation,
+    completionConfig: DirectoryFindCompletionConfig,
+    exclude:          Set[String],
+    continueFrom:     Option[String],
+  )(
+    implicit
+    clock: Clock,
+  ): IO[DirectoryFindResults] =
+    for {
+      listObjsReq <- createListObjectsRequest(bucketAndPrefix, continueFrom)
+      paginator   <- IO(s3Client.listObjectsV2Paginator(listObjsReq))
+      prefixInfo  <- extractPrefixesFromResponse(paginator, exclude, completionConfig)
+      flattened   <- flattenPrefixes(bucketAndPrefix, prefixInfo.partitions, completionConfig, exclude)
+    } yield {
+      prefixInfo match {
+        case CompletedDirectoryFindResults(_)        => CompletedDirectoryFindResults(flattened)
+        case a @ PausedDirectoryFindResults(_, _, _) => a.copy(flattened)
+      }
+    }
+
+  private def flattenPrefixes(
+    bucketAndPrefix:  RemoteS3RootLocation,
+    prefixes:         Set[String],
+    completionConfig: DirectoryFindCompletionConfig,
+    exclude:          Set[String],
+  )(
+    implicit
+    clock: Clock,
+  ): IO[Set[String]] =
+    if (completionConfig.levelsToRecurse == 0) IO.pure(prefixes)
+    else
+      for {
+        ioPrefixes <- IO.pure(prefixes)
+        bap        <- IO.pure(ioPrefixes.map(bucketAndPrefix.fromRoot))
+        recursiveResult <- bap
+          .map(
+            findDirectories(
+              _,
+              completionConfig.copy(levelsToRecurse = completionConfig.levelsToRecurse - 1),
+              exclude,
+              Option.empty,
+            ).map(_.partitions),
+          ).toList.sequence.map {
+            listSetString: List[Set[String]] =>
+              listSetString.toSet.flatten
+          }
+      } yield {
+        recursiveResult.headOption.fold(ioPrefixes)(_ => recursiveResult)
+      }
+
+  private def createListObjectsRequest(
+    bucketAndPrefix: RemoteS3RootLocation,
+    lastFound:       Option[String],
+  ): IO[ListObjectsV2Request] =
+    IO.pure {
+      val builder = ListObjectsV2Request
+        .builder()
+        .maxKeys(1000)
+        .bucket(bucketAndPrefix.bucket)
+        .delimiter("/")
+        .startAfter(lastFound.orNull)
+      bucketAndPrefix.prefix.map(addTrailingSlash).foreach(builder.prefix)
+      builder.build()
+    }
+
+  private def extractPrefixesFromResponse(
+    listObjectsV2Response: ListObjectsV2Iterable,
+    exclude:               Set[String],
+    completionConfig:      DirectoryFindCompletionConfig,
+  ): IO[DirectoryFindResults] =
+    IO {
+      val rtn: mutable.Set[String] = mutable.Set()
+      val resp = listObjectsV2Response
```

Comment:
If lostObjectsV2Response is converted to a Scala iterator you can just use 
```
foldLeft(Set.empty){ case (set, item)=> 

...
set ++ commonPRefixesFiltered
}
```

Reply:
I tried this however this doesn't quite here because we can't loop discriminately through the whole of the list as the iterator hides paging functionality (calling more pages, which may not be desired depending on the configuration).  The way it is implemented we are able to pause the process before we iterate through every item.

I would like a better way of doing this, however as I appreciate it is not too elegant.